acquia/cohesion
# Sync file metadata changed value from YAML

### Motivation

When syncing shared Site Studio configurations between multisite instances on same codebase, there are many sitestudio_file_metadata files that are eternally out of sync.

The changes in these files show a different value for the "changed" property.

```diff
--- a/config/site_studio/sitestudio_file_metadata.AAAAAAAA-BBBB-CCCC-DDDD-EEEEEEEEEEEE.yml
+++ b/config/site_studio/sitestudio_file_metadata.AAAAAAAA-BBBB-CCCC-DDDD-EEEEEEEEEEEE.yml
@@ -6,4 +6,4 @@ filemime: image/png
filesize: '1639'
status: '1'
created: '1659012010'
-changed: '1682768296'
+changed: '1679090222'
```

If I commit this change, then the value becomes out of sync in other multisite databases upon package export.

```bash
$ ddev drush @site1.ddev cex -y && ddev drush @site1.ddev sitestudio:package:export -y
$ git status
On branch develop
Your branch is up to date with 'origin/develop'.
nothing to commit, working tree clean

$ ddev drush @site2.ddev cim -y && ddev drush @site2.ddev sitestudio:package:import -y
$ ddev drush @site2.ddev cex -y && ddev drush @site2.ddev sitestudio:package:export -y
$ git status
On branch develop
Your branch is up to date with 'origin/develop'.
Changes not staged for commit:
(use "git add <file>..." to update what will be committed)
(use "git restore <file>..." to discard changes in working directory)
modified: config/site_studio/sitestudio_file_metadata.AAAAAAAA-BBBB-CCCC-DDDD-EEEEEEEEEEEE.yml
````

The problem here boils down to the fact that each multisite instance imports the shared static file assets, such as new image files added to Site Studio components, at different times, but the `changed` timestamp in the file managed table is always set to the current `time()` any time the config is imported after a file change was detected instead of the `changed` timestamp in the YAML file itself.

_Fixes Acquia Support Case Number 00971770_ 

### Proposed changes

Ensure the `changed` timestamp in the file metadata YAML is taken into account when calculating whether a file has changed and if so, ensure that the new `changed` timestamp in the YML file is in fact synced into the database during Site Studio sync import.

### Testing steps

1. Setup an Acquia CMS instance using the standard docroot/sites/default/settings.php
2. Export all configuration to config/sync/ and config/site_studio/ folders.
3. Convert the codebase into a Drupal multisite setup, by adding a second site instance at docroot/sites/site2/ folder. Ensure settings.php points to the shared config/sync and config/site_studio folders.
4. Ensure site1 and site2 are in sync with codebase (drupal and sitestudio packages).
5. Change an image on the filesystem at config/site_studio/someimage.png 
6. Import config on site1: `drush --uri=site1 sitestudio:package:import`.
7. Export config on site1: `drush --uri=site1 sitestudio:package:export` to get the necessary YAML file changes, typically including a different filesize and changed timstamp.
8. Commit the changes.
9. Import config on site2: `drush --uri=site2 sitestudio:package:import -y`
10. Export config on site2: `drush --uri=site2 sitestudio:package:export -y` 
11. Do a git diff to see that the changed timestamp for the file in question is different.
